### PR TITLE
Entry node reliability score check

### LINKED ISSUE #126: Select reliable nodes actively
### Description

We should check the reliability score of nodes to make sure we are selecting the best nodes possible.
This should be achieved by actively checking, before or after a request, the reliability score and if it's bad, we should select a more reliable node.

### Relevant issues

- [x] Check before or after a request for the reliability score of the entry and exit node.
- [x] Select a different node if it's dishonest or it has a bad score.
- [x] All test pass

### Definition of DONE

We should have a PR with the implementation linked to this issue.

### PR DESCRIPTION
**Pull request by ItsLuized**
## We should actively check the realiability score for the entry node selected.
If the score is lower than 0.7 and it's not a fresh node, we should select a new node.

Fixes #126 #153

### CODE REVIEW

**Review by nionis — CHANGES_REQUESTED** [review]
- tests are failing
- pls merge connect-ecosystem branch in this branch as it is not up to sync

File: packages/sdk/src/index.ts [hunk 1]
```diff
@@ -279,13 +279,21 @@ export default class SDK {
...
    * @param provider
    * @param body
    * @returns Request
    */
   public createRequest(provider: string, body: string): Request {
     if (!this.isReady) throw Error("SDK not ready to create requests");
+    let entryNodeScore: number = this.reliabilityScore.getScore(
+      this.entryNode!.peerId
+    );
+    let counter: number = 0;
```
Comment (nionis) [inline packages/sdk/src/index.ts:291]: what is the use of the counter?

Comment (ItsLuized) [reply packages/sdk/src/index.ts:291]: It's a limit of how many new entry nodes we would search for (the idea is to not create too much latency when creating the request)

Comment (nionis) [reply packages/sdk/src/index.ts:291]: 1. what happens when u reach 3?
2. this is not awaited so it wouldn't create latency

In general I don't think the implementation here is correct - I would rather update the discovery platform API to take in the `request/entry-node` an exclude list which would contain the previously bad entry node and give u a new entry node

File: .gitignore [hunk 2]
```diff
@@ -35,3 +35,5 @@ yarn-error.log*
 .turbo
 
 tmp
+
+.dockerignore
```
Comment (nionis) [inline .gitignore:39]: I think Martins suggested we keep this @sitilge

File: apps/discovery-platform/src/db/index.spec.ts [hunk 3]
```diff
@@ -100,7 +100,9 @@ describe("test db functions", function () {
     await db.saveRegisteredNode(dbInstance, secondNode);
     await db.saveRegisteredNode(dbInstance, createMockNode());
 
-    const notExitNodes = await db.getNonExitNodes(dbInstance);
+    const notExitNodes = await db.getRegisteredNodesWithFilters(dbInstance, {
```
Comment (nionis) [inline apps/discovery-platform/src/db/index.spec.ts:103]: why not turn `getRegisteredNodesWithFilters` to `getRegisteredNodes` and make the filters optional argument?

File: packages/sdk/src/index.ts [hunk 4]
```diff
@@ -279,13 +283,24 @@ export default class SDK {
...
    */
   public createRequest(provider: string, body: string): Request {
     if (!this.isReady) throw Error("SDK not ready to create requests");
+    let entryNodeScore: number = this.reliabilityScore.getScore(
+      this.entryNode!.peerId
+    );
+    const exclusionList: string[] = [];
+    while (entryNodeScore < 0.7 && entryNodeScore !== 0.2) {
+      exclusionList.push(this.entryNode!.peerId);
+      this.selectEntryNode(
```
Comment (nionis) [inline packages/sdk/src/index.ts:298]: shouldn't this be awaited?

Comment (nionis) [reply packages/sdk/src/index.ts:298]: also I would add some logs, this part seems important

File: packages/sdk/src/index.ts [hunk 5]
```diff
@@ -279,13 +283,24 @@ export default class SDK {
...
+        this.ops.discoveryPlatformApiEndpoint,
+        exclusionList
+      );
+      entryNodeScore = this.reliabilityScore.getScore(this.entryNode!.peerId);
```
Comment (nionis) [inline packages/sdk/src/index.ts:302]: re-assignment here is not needed

File: packages/sdk/src/reliability-score.ts [hunk 6]
```diff
@@ -158,9 +158,9 @@ export default class ReliabilityScore {
         this.score.set(peerId, score);
         log.normal("node %s has a %s realiability score", peerId, score);
       }
-      return this.score.get(peerId) || 0;
+      return this.score.get(peerId)!;
     } else {
-      return 0;
+      return 0.2;
```
Comment (nionis) [inline packages/sdk/src/reliability-score.ts:163]: lets make this `0.2` a constant top of this file

**Review by nionis — CHANGES_REQUESTED** [review]

File: packages/sdk/src/index.ts (hunk 4 above)
Comment (ItsLuized) [reply packages/sdk/src/index.ts:298]: Logs added and await for selectEntryNode added as well

**Review by nionis — APPROVED** [review]

**Pull request merged**